src/libutil/ustring.cpp

```diff
@@ -433,12 +428,81 @@ ustring::make_unique(string_view strref)
         strref = string_view("", 0);
 
     size_t hash = Strutil::strhash(strref);
+    // This line, if uncommented, lets you force lots of hash collisions:
+    // hash &= ~size_t(0xffffff);
 
+#if !PREVENT_HASH_COLLISIONS
     // Check the ustring table to see if this string already exists.  If so,
     // construct from its canonical representation.
     // NOTE: all locking is performed internally to the table implementation
     const char* result = table.lookup(strref, hash);
     return result ? result : table.insert(strref, hash);
+
+#else
+    // Check the ustring table to see if this string already exists with the
+    // default hash. If so, we're done. This is by far the common case --
+    // most lookups already exist in the table, and hash collisions are
+    // extremely rare.
+    const char* result = table.lookup(strref, hash);
```

Comment:
I really don't mind if checking for collisions, or calling OIIO::ustring::hash_collisions is slow, since it should be expected that you won't do this "every time". It should be expected to be a debugging tool, not a regular part of execution. As such, I agree with what @fpsunflower is saying above. There doesn't seem to be a need to increase memory and runtime just to make an operation you don't want to run every time, work slightly better.

Reply:
@fpsunflower Sorry for the delay. I've put this through its paces in thread-wedged torture tests of ustring_test, and there is no consistent detectable performance hit, and also I ran the usual SPI production tests and if there is any perf difference at all, it's within the 1% timing noise slop we typically see. I just can't demonstrate it. It's true that there are more locks (only when each unique ustring is created the first time, it's the usual one lock as before when you are creating a ustring from characters and that character sequence has been encountered before). The lock on the reverse map is not global, it's heavily sharded, so two threads needing the same lock should be rare, even in times of heavy multithreaded string creation.

@etheory, The speedup of hash_collisions is a nice side benefit, but not the purpose of the patch. Remember, what we are trying to do is ensure 100% certainty that even if there is a collision on the initial constexpr hash guess, the strings will end up with rehashed values that are guaranteed to be unique.

If this still is worrying, or if the extra memory is more of a concern than the time, or if this starts to show up in anybody's profile in the future, we could turn it into something that is disabled by default and can be turned on at runtime. But I am dubious that the very strange errors you could get (but probably never will) from a collision would be very hard for users to understand and know to turn on the double checking option.